packages/grid_client/tests/modules/zmachine.test.ts

```diff
@@ -29,36 +35,136 @@ beforeAll(() => {
   zmachine.mounts = [disks];
   zmachine.entrypoint = "/sbin/zinit init";
   zmachine.compute_capacity = computeCapacity;
-  zmachine.env = {};
+  zmachine.env = { key: "value" };
   zmachine.corex = false;
-  zmachine.gpu = [];
+  zmachine.gpu = ["AMD", "NIVIDIA"];
 });
 
 describe("Zmachine Class Tests", () => {
   it("should create a valid Zmachine instance", () => {
     expect(zmachine).toBeInstanceOf(Zmachine);
   });
 
+  it("should correctly serialize and deserialize a Zmachine instance", () => {
+    const serialized = JSON.stringify(zmachine);
+    const deserialized = plainToClass(Zmachine, JSON.parse(serialized));
+
+    expect(deserialized).toBeInstanceOf(Zmachine);
+    expect(deserialized.challenge()).toBe(zmachine.challenge());
+  });
+  it("should correctly handle env vars", () => {
+    const challenge = zmachine.challenge();
+
+    expect(challenge).toContain("key=value");
+  });
+
   it("should correctly compute the challenge string", () => {
     const expectedChallenge =
-      "https://hub.grid.tf/tf-official-vms/ubuntu-22.04.flist" +
-      "10.249.0.0/16" +
-      "true" +
-      "znetwork" +
-      "10.20.2.22" +
-      "14748364815" +
-      "zdisk" +
-      "/mnt/data" +
-      "/sbin/zinit init";
+      zmachine.flist +
+      network.challenge() +
+      zmachine.size +
+      computeCapacity.challenge() +
+      zmachine.mounts[0].challenge() +
+      zmachine.entrypoint +
+      JSON.stringify(zmachine.env)
+        .replace(/[{"}"]/g, "")
+        .replace(":", "=") +
+      zmachine.gpu?.toString().replace(",", "");
 
     expect(zmachine.challenge()).toBe(expectedChallenge);
+    expect(zmachine.challenge()).toContain("key=value");
   });
 
-  it("should fail validation for missing required fields", () => {
-    const result = () => {
+  it("should correctly handle the gpu array", () => {
+    expect(zmachine.gpu).toContain("NIVIDIA");
+
+    zmachine.gpu = [];
+
+    expect(zmachine.challenge()).not.toContain("NIVIDIA");
+
+    zmachine.gpu = ["NIVIDIA", "AMD"];
+
+    expect(zmachine.challenge()).toContain("NIVIDIA");
+    expect(zmachine.challenge()).toContain("AMD");
+  });
+
+  it("should fail validation for entering invalid flist", () => {
+    const invalidFlist = () => {
       zmachine.flist = "";
-      zmachine.entrypoint = "";
     };
+
+    expect(invalidFlist).toThrow();
+  });
+
+  it("should fail validation for entering invalid entrypoint", () => {
+    const invalidEntrypoint = () => (zmachine.entrypoint = undefined as any);
+
+    expect(invalidEntrypoint).toThrow();
+  });
+
+  it("should fail validation for entering invalid size", () => {
+    const invalidSize = () => (zmachine.size = 10 * 1024 ** 5);
+
+    expect(invalidSize).toThrow();
+  });
```

Comment:
you can try negative numbers or fractions

Reply:
I added more assertions and fixed the validation for flist & size in zmachine.